File: cloud/control-plane/internal/contract/telemetry_test.go
package contract_test

import (
	"testing"

	"google.golang.org/protobuf/proto"

	"github.com/orvolt/orvolt/cloud/control-plane/internal/contract"
	evsev1 "github.com/orvolt/orvolt/contracts/gen/go/orvolt/telemetry/evse/v1"
)

func TestDecodeChargingTelemetry(t *testing.T) {
	payload, err := proto.Marshal(&evsev1.ChargingTelemetry{
		StationId:    "station-1",
		ConnectorId:  "1",
		TimestampMs:  1_700_000_000_000,
		Voltage:      400,
		Current:      75,
		PowerKw:      30,
		EnergyKwh:    20,
		Soc:          50,
		TemperatureC: 33,
		State:        evsev1.ChargingState_CHARGING,
		Edge:         &evsev1.EdgeMetadata{EdgeId: "edge-1", SiteId: "site-1", ReceivedAtMs: 1_700_000_000_001},
	})
	if err != nil {
		t.Fatal(err)
	}

	telemetry, err := contract.DecodeChargingTelemetry(payload)
	if err != nil {
		t.Fatal(err)
	}
	if telemetry.StationID != "station-1" || telemetry.State != "CHARGING" || telemetry.SiteID != "site-1" {
		t.Fatalf("unexpected decoded telemetry: %#v", telemetry)
	}
}

File: simulator/evse-simulator/tests/telemetry_test.cpp
#include "orvolt/simulator/telemetry.hpp"

#include <cassert>

int main() {
  orvolt::simulator::TelemetryGenerator generator{"test-station", "1", 7};
  bool observed_charging = false;
  for (int index = 0; index < 50; ++index) {
    const auto telemetry = generator.next();
    assert(telemetry.station_id == "test-station");
    assert(telemetry.connector_id == "1");
    assert(telemetry.timestamp_ms > 0);
    assert(telemetry.soc >= 0.0 && telemetry.soc <= 100.0);
    assert(telemetry.temperature_c >= 29.0 && telemetry.temperature_c <= 38.0);
    if (telemetry.state == orvolt::simulator::ChargingState::Charging) {
      observed_charging = true;
      assert(telemetry.voltage >= 390.0 && telemetry.voltage <= 410.0);
      assert(telemetry.current >= 68.0 && telemetry.current <= 78.0);
      assert(telemetry.power_kw > 0.0);
    }
    assert(orvolt::simulator::to_json(telemetry).find("\"station_id\":\"test-station\"") != std::string::npos);
  }
  assert(observed_charging);
}
